spec/micro/README.md

```diff
@@ -0,0 +1,770 @@
+
+    // assert dest channel is sourceChannel's counterparty channel identifier
+    counterparty = getCounterparty(packet.destChannel)
+    assert(packet.sourceChannel == counterparty.channelId)
+   
+    // assert dest port is sourcePort's counterparty port identifier
+    assert(packet.destPort == ports[packet.sourcePort])
+
+    // verify we sent the packet and haven't cleared it out yet
+    assert(provableStore.get(packetCommitmentPath(packet.sourcePort, packet.sourceChannel, packet.sequence))
+           === hash(hash(packet.data), packet.timeoutHeight, packet.timeoutTimestamp))
+
+    // get the timestamp from the final consensus state in the channel path
+    var proofTimestamp
+    proofTimestamp = client.getTimestampAtHeight(proofHeight)
+    assert(err != nil)
+
+    // check that timeout height or timeout timestamp has passed on the other end
+    asert(
+      (packet.timeoutHeight > 0 && proofHeight >= packet.timeoutHeight) ||
+      (packet.timeoutTimestamp > 0 && proofTimestamp >= packet.timeoutTimestamp))
+
+    receiptPath = packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence)
+    merklePath = applyPrefix(counterparty.keyPrefix, receiptPath)
+    assert(client.verifyNonMembership(
+        proofHeight
+        0, 0,
+        proof,
+        merklePath
+    ))
+
+    cbs = router.callbacks[packet.sourcePort]
+    cbs.OnTimeoutPacket(packet, relayer)
+
+    channelStore.delete(packetCommitmentPath(packet.sourcePort, packet.sourceChannel, packet.sequence))
+}
+```
+
+### Correctness
+
+Claim: If the clients are setup correctly, then a chain can always verify packet flow messages sent by a valid counterparty.
+
+If the clients are correct, then they can verify any key/value membership proof as well as a key non-membership proof.
+
+All packet flow message (SendPacket, RecvPacket, and TimeoutPacket) are sent with the full packet. The packet contains both sender and receiver identifiers. Thus on packet flow messages sent to the receiver (RecvPacket), we use the receiver identifier in the packet to retrieve our local client and the source identifier to determine which path the sender stored the packet under. We can thus use our retrieved client to verify a key/value membership proof to validate that the packet was sent by the counterparty.
+
+Similarly, for packet flow messages sent to the sender (AcknowledgePacket, TimeoutPacket); the packet is provided again. This time, we use the sender identifier to retrieve the local client and the destination identifier to determine the key path that the receiver must have written to when it received the packet. We can thus use our retrieved client to verify a key/value membership proof to validate that the packet was sent by the counterparty. In the case of timeout, if the packet receipt wasn't written to the receipt path determined by the destination identifier this can be verified by our retrieved client using the key nonmembership proof.
+
+### Soundness
```

Comment:
Note that, potentially, in the future we could enforce client identifiers global uniqueness. For now I agree that  the packet is committed key path is a strong guarantee.